addons/hr_holidays/models/hr_holidays.py

```diff
@@ -0,0 +1,557 @@
+        help="Choose 'Leave Request' if someone wants to take an off-day. "
+             "\nChoose 'Allocation Request' if you want to increase the number of leaves available for someone")
+    parent_id = fields.Many2one('hr.holidays', string='Parent')
+    linked_request_ids = fields.One2many('hr.holidays', 'parent_id', string='Linked Requests')
+    department_id = fields.Many2one('hr.department', related='employee_id.department_id', string='Department', readonly=True, store=True)
+    category_id = fields.Many2one('hr.employee.category', string='Employee Tag', readonly=True,
+        states={'draft': [('readonly', False)], 'confirm': [('readonly', False)]}, help='Category of Employee')
+    holiday_type = fields.Selection([
+        ('employee', 'By Employee'),
+        ('category', 'By Employee Tag')
+    ], string='Allocation Mode', readonly=True, required=True, default='employee',
+        states={'draft': [('readonly', False)], 'confirm': [('readonly', False)]},
+        help='By Employee: Allocation/Request for individual Employee, By Employee Tag: Allocation/Request for group of employees in category')
+    manager_id2 = fields.Many2one('hr.employee', string='Second Approval', readonly=True, copy=False,
+        help='This area is automaticly filled by the user who validate the leave with second level (If Leave type need second validation)')
+    double_validation = fields.Boolean('Apply Double Validation', related='holiday_status_id.double_validation')
+    can_reset = fields.Boolean('Can reset', compute='_compute_can_reset')
+
+    @api.multi
+    @api.depends('number_of_days_temp', 'type')
+    def _compute_number_of_days(self):
+        for holiday in self:
+            if holiday.type == 'remove':
+                holiday.number_of_days = -holiday.number_of_days_temp
+            else:
+                holiday.number_of_days = holiday.number_of_days_temp
+
+    @api.multi
+    def _compute_can_reset(self):
+        """ User can reset a leave request if it is its own leave request
+            or if he is an Hr Manager.
+        """
+        user = self.env.user
+        group_hr_manager = self.env.ref('base.group_hr_manager')
+        for holiday in self:
+            if group_hr_manager in user.groups_id or holiday.employee_id and holiday.employee_id.user_id == user:
+                holiday.can_reset = True
+
+    @api.constrains('date_from', 'date_to')
+    def _check_date(self):
+        for holiday in self:
+            domain = [
+                ('date_from', '<=', holiday.date_to),
+                ('date_to', '>=', holiday.date_from),
+                ('employee_id', '=', holiday.employee_id.id),
+                ('id', '!=', holiday.id),
+                ('state', 'not in', ['cancel', 'refuse']),
+            ]
+            nholidays = self.search_count(domain)
+            if nholidays:
+                raise ValidationError(_('You can not have 2 leaves that overlaps on same day!'))
```

Comment:
Could be done with an SQL constraint (using range types), and the validation message seems incorrect.